zookeeper-server/src/main/java/org/apache/zookeeper/server/auth/znode/groupacl/X509ZNodeGroupAclProvider.java

```diff
@@ -0,0 +1,156 @@
+ *    need to be read by multiple identities from multiple domains.
+ *  3.Multiple creators, one all-permission accessor - ZNodes, created by multiple identities from multiple domains,
+ *    need to be read/update/deleted by an identity from a different domain.
+ * Optional features include:
+ *    set client Id as znode ACL when creating nodes,
+ *    add additional public read ACL to znodes based on path.
+ */
+public class X509ZNodeGroupAclProvider extends ServerAuthenticationProvider {
+
+  private static final Logger LOG = LoggerFactory.getLogger(X509ZNodeGroupAclProvider.class);
+  private final String logStrPrefix = this.getClass().getName() + ":: ";
+  private final X509TrustManager trustManager;
+  private final X509KeyManager keyManager;
+  static final String ZOOKEEPER_ZNODEGROUPACL_SUPERUSER = "zookeeper.znodeGroupAcl.superUser";
+
+  public X509ZNodeGroupAclProvider() {
+    ZKConfig config = new ZKConfig();
+    this.keyManager = X509AuthenticationUtil.createKeyManager(config);
+    this.trustManager = X509AuthenticationUtil.createTrustManager(config);
+  }
+
+  public X509ZNodeGroupAclProvider(X509TrustManager trustManager, X509KeyManager keyManager) {
+    this.trustManager = trustManager;
+    this.keyManager = keyManager;
+  }
+
+  @Override
+  public KeeperException.Code handleAuthentication(ServerObjs serverObjs, byte[] authData) {
+    ServerCnxn cnxn = serverObjs.getCnxn();
+    X509Certificate clientCert;
+    try {
+      clientCert = X509AuthenticationUtil.getAndAuthenticateClientCert(cnxn, trustManager);
+    } catch (KeeperException.AuthFailedException e) {
+      return KeeperException.Code.AUTHFAILED;
+    }
+
+    // Extract URI from certificate
+    String uri;
+    try {
+      uri = X509AuthenticationUtil.getClientId(clientCert);
+    } catch (Exception e) {
+      // Failed to extract URI from certificate
+      LOG.error(logStrPrefix + "Failed to extract URI from certificate for session 0x{}",
+          Long.toHexString(cnxn.getSessionId()), e);
+      return KeeperException.Code.OK;
+    }
+
+    // User belongs to super user group
+    if (uri.equals(System.getProperty(ZOOKEEPER_ZNODEGROUPACL_SUPERUSER))) {
+      cnxn.addAuthInfo(new Id("super", uri));
+      LOG.info("Authenticated Id '{}' as super user", uri);
+      return KeeperException.Code.OK;
+    }
+
+    // Get authorized domain names for client
+    ClientUriDomainMappingHelper uriDomainMappingHelper =
+        new ZkClientUriDomainMappingHelper(serverObjs.getZks());
+    Set<String> domains = uriDomainMappingHelper.getDomains(uri);
+    if (domains.isEmpty()) {
+      // If no domain name is found, use URI as domain name
+      domains = new HashSet<>();
+      domains.add(uri);
+    }
+
+    Set<String> superUserDomainNames = ZNodeGroupAclProperties.getInstance().getSuperUserDomainNames();
+    for (String domain : domains) {
+      // Grant cross domain components super user privilege
```

Comment:
If we're going to use "cross domain components", should we first define it in the JavaDoc?

Reply:
Let me correct this statement here. Super user domains should not be limited to "cross domain components". "Cross domain components" is an internal term and should not be mentioned here.